### ISSUE TITLE
Angular 6 - ngServe & loadChildren causes SO 

### ISSUE THREAD

**Issue by ludydoo** [angular/angular#24539]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report 
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
I made a clean app with angular (ng new). I want to use lazy loading because some third party libraries are huge. In the AppRouterModule, I added a route 

```
 {
    path: 'about',
    component: AboutComponent,
    loadChildren: './../dev-express/dev-express.module#DevExpressModule'
  }
```

And in the module definition of the RouterModule : I added
`imports: [R.forRoot(routes, {preloadingStrategy: PreloadAllModules})],`

The DevExpressModule is as follows : 

```
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DevExtremeModule } from 'devextreme-angular';

@NgModule({
  imports: [
    CommonModule,
    DevExtremeModule
  ],
  exports: [
    DevExtremeModule
  ],
  declarations: []
})
export class DevExpressModule { }

```

Now, when I run ng serve or ng-build, I get the expected separate chunks : 

![capture](https://user-images.githubusercontent.com/38968/41471115-ae4e65f4-7080-11e8-9d00-7a6409cdf881.JPG)

But if I run this in the browser (either ng serve or ng dist then serve through server), I get a stack overflow exception : 

![capture2](https://user-images.githubusercontent.com/38968/41471154-d42eec4e-7080-11e8-8750-7a71b919fc97.JPG)

When I don't use loadChildren, and I include the Module directly in the "includes", I get no error at runtime.

## Expected behavior

Not having a stack overflow when using loadChildren

## Minimal reproduction of the problem with instructions

https://stackblitz.com/edit/angular-gitter-2dbbh1

## What is the motivation / use case for changing the behavior?

Pretty obvious?

## Environment

<pre><code>
Angular version: 6.0.5

Browser:
- [x ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.11.3
- Platform:  Win 10 Pro x64

Others:

</code></pre>

**Comment by jasonaden** [comment]
@ludydoo Is this something that has worked in the past? Is there a version of Angular this works with? Trying to find out if this is a regression, and where it broke, or if this has been existing for a while.

Also, can you replicate this with a module other than DevExtremeModule? Perhaps there is an issue with that module specifically. Can you change the repro to be simple and not pull in any external libraries?

**Comment by ludydoo** [comment]
Yes I resolved the issue little bit after posting this. If I remember correctly, the problem was that a module cannot *not* declare at least a component. In that case, I think the router looped around to find a suitable component but could not find one. My motivation for having a module that only imports other module was that I wanted to preload at a certain point in time some dependencies (in this case, DevExpress). So by adding a dummy component to the module, the router did not complain anymore.

**Comment by scott-ho** [comment]
There is some problem inside route_preloader

see here: https://github.com/angular/angular/blob/c95437f15df93e2ca735ebf015acb20be597bb6a/packages/router/src/router_preloader.ts#L121

There exists some cases that the `res` will be empty, which causes the stream to be hung

**Comment by deskenazi** [comment]
Just want to add that we experienced a similar problem with a different cause. 
In our case the modules loaded perfectly when lazy loaded but when we switched to PreloadAllModules the app would hang and if you waited long enought you would get the stack overflow. 
After debugging the Angular code around  processRoutes() we discovered that in our case the recursion of processRoutes never ended because we had Cirular Module dependencies. Once we suspected this we used the Augury routes visualization to discover the specific problem modules and corrected the issue.

**Comment by jasonaden** [comment]
Closing this issue based on the resolution above. We can open up a new issue around detecting circular references, but the specific issue mentioned here was resolved by the OP.
